src/System.Management.Automation/engine/lang/parserutils.cs

```diff
@@ -1347,35 +1346,39 @@ internal static object CompareOperators(ExecutionContext context, IScriptExtent
         }
 
         /// <summary>
-        /// Cache regular expressions...
+        /// Cache regular expressions.
         /// </summary>
         /// <param name="patternString">The string to find the pattern for.</param>
-        /// <param name="options">The options used to create the regex...</param>
-        /// <returns>A case-insensitive Regex...</returns>
+        /// <param name="options">The options used to create the regex.</param>
+        /// <returns>New or cached Regex.</returns>
         internal static Regex NewRegex(string patternString, RegexOptions options)
         {
-            if (options != RegexOptions.IgnoreCase)
-                return new Regex(patternString, options);
-
-            lock (s_regexCache)
+            var subordinateRegexCache = s_regexCache.GetOrAdd(options, s_subordinateRegexCacheCreationDelegate);
+            if (subordinateRegexCache.TryGetValue(patternString, out Regex result))
             {
-                Regex result;
-                if (s_regexCache.TryGetValue(patternString, out result))
-                {
-                    return result;
-                }
-                else
+                return result;
+            }
+            else
+            {
+                return subordinateRegexCache.GetOrAdd(patternString, key =>
```

Comment:
Have this lambda expression in this method will result in an instance of the helper class to be created every time `NewRegex` gets called, even if the query can be served by the cache.

This should be changed to the following:
```c#
else
{
    if (subordinateRegexCache.Count > MaxRegexCache)
    {
        subordinateRegexCache.Clear();
    }

    var regex = new Regex(patternString, options);
    return subordinateRegexCache.GetOrAdd(patternString, regex);
}
```
Note that, the `ValueFactory` delegate is called outside the locks and thus it's not atomic, meaning there is no difference clearing the dictionary this way from doing it in the `ValueFactory` delegate.